### PR TITLE
DIfferent types of forms, everytime in a railsy-er way than the one before

### PR DESCRIPTION
**Pull request by YoseptF**
In this exercise we made some examples of how to make a form inside a rails app, there's a new one (form_with), but it is not covered in this version of the book.

## Versions
- Basic Html form
  - It's the most basic (as the name suggests it), and it relies only on Html to be created, so you need to add encryption so that rails can read it.
- form_tag
  - It's the first rails way, it automatically adds the auth input but otherwise, it's still pretty much like making a pure Html form
- form_for
  - It's the most railsy way to do forms (in rails 5), it automatically sends the authentication, and also nest the field inputs into a hash before sending it.

---

Here's a preview of our form:
image

---

Once again, thanks for the review, we are happy to hear your comments. :tada:

### CODE REVIEW

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -0,0 +1,38 @@
+class UsersController < ApplicationController
+  def new
+    @user = User.new
+  end
+
+  def create
+    @user = User.new(user_params)
+    # @user = User.new(username:params[:username], email:params[:email], password:params[:password])
```
Comment (maelfosso) [inline app/controllers/users_controller.rb]: Please, remove comments

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+    if @user.save
+      render json: { params: params, status: 'saved' }, status: :ok
```
Comment (maelfosso) [inline app/controllers/users_controller.rb]: As indicated [here](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/forms/user_create.png), `redirect_to new_user_path` if user was saved

File: app/controllers/users_controller.rb [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+  end
+
+  def edit
+    @user = User.find(params[:id])
+  end
+
+  def update
+    @user = User.find(params[:id])
+    if @user.update(user_params)
+      render json: @user
```
Comment (maelfosso) [inline app/controllers/users_controller.rb]: `redirect_to edit_user_path` if user was saved

File: app/controllers/users_controller.rb [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+  def create
+    @user = User.new(user_params)
+    # @user = User.new(username:params[:username], email:params[:email], password:params[:password])
+
+    if @user.save
+      render json: { params: params, status: 'saved' }, status: :ok
+    else
+      render 'new'
+      # render json: @user.errors.full_messages
```
Comment (maelfosso) [inline app/controllers/users_controller.rb]: Remove all comments

File: app/views/users/new.html.erb [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+  <%= label_tag :username, "Username" %>
+  <%= text_field_tag :username, nil, placeholder:"Enter your desired Username" %>
+  <%= label_tag :email, "Email" %>
+  <%= text_field_tag :email, nil, placeholder:"Enter a valid Email" %>
+  <%= label_tag :password, "Password" %>
+  <%= password_field_tag :password, nil %>
+  <%= submit_tag :submit %>
+<% end %>
+-->
+<%= render 'shared/form' %>
```
Comment (maelfosso) [inline app/views/users/new.html.erb]: For all the form, please, use `users_path` helper for action

**Review by maelfosso — CHANGES_REQUESTED** [review]
# Changes required :red_circle: 

Hello @YoseptF  @NewIncome 

I'm a fan of your work!!! Well job done !!!
Before moving forward, there are some changes to do

1. Please, remove all the comments.

2. Read the comments left below.

Kindly make the suggested changes and submit a new code review request.

Thanks
Mael FOSSO

**Comment by YoseptF** [comment]
Hi @maelfosso,

After the review, we made some changes in the code and now:

- [X] It uses user_path in all the form to select the URL where the action is gonna take place
  - We saw in your comment that we needed to use `users_path`, but in our routes, that's a get route pointing towards #creat, we think you meant `user_path`, which points towards the #edit route

image

image

- [X] The user is now redirected to the correct route after the new user is created
  - they also get a notification about the current status of the new user, so that is clear
image


- [X] The user is now redirected to the correct route after a user is successfully edited
  - they also get a notification about the current status of the new user, so that is clear
image
- [X] Finally, we deleted all the unnecessary comments from the code (except for the forms we are not using, but we think that's kinda part of the review). Sadly there's no image for that :cry: 

---

As always thanks for your time, and we are waiting for your comments :tada:

**Review by uche-inyama — APPROVED** [review]
## Approved

Hi team!

Nice, you have implemented the required changes.

However, there's an observation: run **```rails routes | grep user```** in your console, you will get the list of routes for user model. You will make the following observations:

- [x] For create action it is **``` user_path ```** 
- [x] Update action it is **``` users_path ```**

 Make the minor change to your form.

I will go ahead and approve your project.   :+1: 

#### ``` Have a great day ``` :slightly_smiling_face: 

``` Uchechukwu Henry Inyama ```

**Pull request merged**